app/modules/ir.c

```diff
@@ -0,0 +1,248 @@
+//#define NODE_DEBUG
+
+#include "module.h"
+#include "lauxlib.h"
+#include "lmem.h"
+#include "platform.h"
+#include "user_interface.h"
+#include <stdint.h>
+#include <string.h>
+#include "gpio.h"
+#include "hw_timer.h"
+
+/* Must be a power of 2 (max 128 without changing types) */
+#define RAWBUF_SIZE 128
+#define RAWBUF_MASK (RAWBUF_SIZE - 1)
+
+#define CB_RAW  0
+#define CB_RC5  1
+
+static task_handle_t tasknumber;
+static ETSTimer timer;
+static struct {
+  int cb[2];
+  uint32_t last_time;
+  uint16_t rawbuf[RAWBUF_SIZE];
+  uint16_t rc5_data;
+  uint8_t rc5_state;
+  uint8_t rawbuf_read;
+  uint8_t rawbuf_write;
+  uint8_t pin, pin_num;
+  uint8_t setup;
+} data;
```

Comment:
I know I'm going to sound like a pest, but: is it possible to avoid globals like this and encapsulate everything needed within a Lua userdata?  Could we even potentially allow for multiple IR rxers at once?

Reply:
Some global data is needed for the interrupt routine (so it can find the structure with the data in). But it would (in theory) be possible to move most of this into dynamically allocated data which would allow multiple IR receivers. I have to admit removing the need for the static `rawbuf` array would be a nice improvement (reduces RAM usage when not in use). I'll have a look at how difficult it would be to do.

Reply:
@ecrips Maybe it isn't the best suggestion, but you can have a look at `softuart` module where I was faced with similar problem. My solution is not ideal (as it cannot read more than one data frame simultaneously) but the method on implementing multiple receiving instances and selecting right one in ISR should be possible to apply in this code.